gramlib/grammar.ml

```diff
@@ -29,7 +29,7 @@ module type S =
       end
     type ('self, 'a) ty_symbol
     type ('self, 'f, 'r) ty_rule
-    type 'a ty_production
+    type ('self, 'a) ty_production
```

Comment:
I am not sure this is the right fix. The question boils down to the exact semantics of a rule of the form `| [ s = SELF -> s ] -> {...}`, i.e. anonymous trees that contain a `SELF` call. I was under the impression that camlp5 was forbidding this kind of situation dynamically, in which case the `s_rules` function should take a universally quantified production rule, and the `ty_production` type should be kept as is. Anybody aware of the semantics of this?

Reply:
To be honest, I just did some kind of "type guided programming" here and was happy to remove the Obj.Magic without modifying the code. I'll have a closer look.

Reply:
There is the same structure in `extend.ml`, you can have a look.

Reply:
Thanks a lot @ppedrot for the explanations!

Looking at camlp5, it appears that it forbids SELF and NEXT in anonymous trees statically, i.e., the preprocessor fails with an error message such as
```
File "f.ml", line 13, characters 17-21:
Parse error: 'SELF' illegal in anonymous entry level
```
Thus, and since I was unable to properly cast the universal type from `extend.ml` to a similar one in `grammar.ml`, I replaced it with type annotations to distinguish rules not containing `SELF` or `ǸEXT` for sure (`norec`) from rules that may contain them (`mayrec`).

Reply:
That's another solution, albeit slightly more invasive. I don't have a particular opinion on the actual technical solution.

Reply:
It's not that I dislike your initial solution but I was unable to build something of type t' = { norec_rule' : 's. 's rule' } given a value of type t = { norec_rule : 's. 's rule } and f : 'a rule -> 'a rule'.

Reply:
I removed some code duplication (not changing the technical solution) but I'm afraid I cannot do any better. So I won't touch this anymore.